crates/llms/src/nql/candle.rs

```diff
@@ -0,0 +1,174 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+     https://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#![allow(clippy::needless_pass_by_value)]
+#![allow(clippy::module_name_repetitions)]
+use async_trait::async_trait;
+use candle_examples::token_output_stream::TokenOutputStream;
+use candle_transformers::{generation::LogitsProcessor, models::quantized_llama::ModelWeights};
+
+use tokenizers::Tokenizer;
+
+use super::{Error as NqlError, Nql, Result};
+use candle_core::{quantized::gguf_file, Tensor};
+
+#[derive()]
+struct InferenceHyperparams {
+    pub to_sample: usize,
+    pub max_seq_len: usize,
+    pub repeat_last_n: usize,
+    pub repeat_penalty: f32,
+    pub device: candle_core::Device,
+    pub seed: u64,
+    pub temperature: f64,
+    pub split_prompt: bool,
+}
+
+impl Default for InferenceHyperparams {
+    fn default() -> Self {
+        Self {
+            to_sample: 300,
+            max_seq_len: 4096,
+            repeat_last_n: 64,
+            repeat_penalty: 1.1,
+            device: candle_core::Device::Cpu,
+            seed: 299_792_458,
+            temperature: 0.8,
+            split_prompt: true,
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct CandleLlama {
+    tknzr: Tokenizer,
+    mdl: ModelWeights,
+}
+
+#[async_trait]
+impl Nql for CandleLlama {
+    async fn run(&mut self, prompt: String) -> Result<Option<String>> {
+        // tknzr.clone() is bad
+        let stream = TokenOutputStream::new(self.tknzr.clone());
+        match Self::perform_inference(prompt, stream, &mut self.mdl.clone()) {
+            Ok(opt_output) => Ok(opt_output),
+            Err(e) => Err(NqlError::FailedToRunModel { e }),
+        }
+    }
+}
+
+impl CandleLlama {
+    pub fn try_new(tokenizer: String, model_weights: String) -> Result<Self> {
```

Comment:
```suggestion
    pub fn try_new(tokenizer: String, model_weights: Path) -> Result<Self> {
```

Could this be a Path or PathBuf?